### PR TITLE
Add support for websocket protocol upgrade.

### PR DESCRIPTION
**Pull request by celiocidral**
A ring handler can detect a websocket protocol upgrade request by inspecting the `Connection` and `Upgrade` headers. Then the ring handler may decide to proceed with the upgrade by returning HTTP status 101 (Switching Protocols) with the appropriate websocket upgrade response headers. Then the ring adapter automatically upgrades the HTTP request to a websocket connection. In this case the ring adapter expects the response map to have a `:ws` key containing the websocket handler.

The motivation here is to allow clients to connect on the same URI path for regular HTTP calls as well as websockets. E.g. GraphQL clients typically connect on the same URI path (e.g. `/graphql`) for running queries and making subscriptions over websocket.

Two unresolved issues in this PR:

* ~~`websocket-upgrade-response?` is case-sensitive with regards to reading the response headers.~~ (done).
* ~~The protocol upgrade support needs to be replicated in `proxy-async-handler`~~ (done).

I'm planning to fix those two issues in case you approve this draft. There might be other issues that escaped me that you'll probably catch. Please let me know.

More information on switching protocols at https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/101

### CODE REVIEW

**Comment by celiocidral** [comment]
I just realized there was already a related open issue: https://github.com/sunng87/ring-jetty9-adapter/issues/41

**Comment by sunng87** [comment]
Thank you! This is a great step to merge our websocket handler with ring handler.

After this patch being merged, I think we can make this a break change to remove legacy websocket handler APIs.

File: examples/rj9a/websocket_upgrade.clj
Comment (sunng87) [inline examples/rj9a/websocket_upgrade.clj:16]: We can make this a built-in helper function for upgrading to websocket connection

Comment (celiocidral) [reply examples/rj9a/websocket_upgrade.clj:16]: Good call! I'll do that.

File: src/ring/adapter/jetty9.clj
Comment (sunng87) [inline src/ring/adapter/jetty9.clj:62]: I wonder if this call is an expensive one? Perhaps we can move this to server start process?

File: examples/rj9a/websocket_upgrade.clj
Comment (celiocidral) [reply examples/rj9a/websocket_upgrade.clj]: I moved those two fns over to the websocket namespace with different names. Please let me know if you think they need better names or live in some other namespace.

**Comment by celiocidral** [comment]
I admit that I felt tempted to make it compatible with the upcoming Ring 2.0 spec, but it requires listeners to implement the `Listener` protocol (which we would have to define first), and it also doesn't include support for ping/pong frames which is something we need in our websockets. Just wondering, do you see any need to have that compatibility in advance, before Ring 2.0 comes out?

File: examples/rj9a/websocket_upgrade.clj
Comment (sunng87) [reply examples/rj9a/websocket_upgrade.clj]: The names looks good to me. Thanks!

**Comment by sunng87** [comment]
@celiocidral We can wait for the final release of Ring 2.0 to support that. Also I am going to support both of Ring's `Listener` protocol and our handler map for websocket. Ping/Pong handling is important feature that I want developer to have control over it.

File: src/ring/adapter/jetty9.clj
Comment (sunng87) [inline src/ring/adapter/jetty9.clj:52]: I wonder if we really need to convert all header names to lower case. This function will be executed on all requests so we'd better keep it as efficient as possible.

Comment (sunng87) [reply src/ring/adapter/jetty9.clj:52]: hmmm... According to https://github.com/ring-clojure/ring/issues/311 we need to deal with the case-issue we accessing headers.

https://github.com/sunng87/ring-jetty9-adapter/blob/81d2b0be1711c5fd798f36cd44f70daf37cef01a/src/ring/adapter/jetty9/common.clj#L18 We have some other case-sensitive header access in this library.

File: src/ring/adapter/jetty9.clj
Comment (sunng87) [inline src/ring/adapter/jetty9.clj:79]: Use our new `ws-upgrade-request?` here?

File: src/ring/adapter/jetty9.clj
Comment (celiocidral) [reply src/ring/adapter/jetty9.clj:52]: > I wonder if we really need to convert all header names to lower case. This function will be executed on all requests so we'd better keep it as efficient as possible.

I'm glad you brought that up because I was hesitant of converting header names at first. The good news is that the `let` form doesn't get evaluated unless the status is 101, which I hope happens less frequently than regular HTTP calls. But anyways, not converting to lower case would be a lot better.

> We have some other case-sensitive header access in this library.

Maybe we could follow that pattern across the board, i.e. case-sensitive, Pascal-kebab case. That means `ws-upgrade-response` would also return header names in that case pattern. What do you think?

File: src/ring/adapter/jetty9.clj
Comment (celiocidral) [reply src/ring/adapter/jetty9.clj:79]: There's a difference between `ws-upgrade-request?` and `websocket-upgrade-response?`. The latter checks the request's http status code, whereas the former doesn't.

File: src/ring/adapter/jetty9.clj
Comment (sunng87) [reply src/ring/adapter/jetty9.clj:52]: `(= 101 status)` does great help here. I suggest to follow the http and ring protocol to keep it case insensitive. I will modify prior issues in separated pull request.

**Review by sunng87 — APPROVED** [review]

**Comment by sunng87** [comment]
Thank you @celiocidral ! I'm going to deprecate `:websockets` option of `run-jetty` and update our README soon.

**Comment by celiocidral** [comment]
Sounds great, thanks!

Just a quick postmortem comment: for comparing header values, I think [String.equalsIgnoreCase](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#equalsIgnoreCase(java.lang.String)) might perform a little bit better than `lower-case`.

**Pull request merged**
